File: src/utils/matchingEngine.js
/**
 * AI Matching Engine - Vector Space Model
 * Calculates compatibility scores using feature vectorization and cosine similarity.
 */

const WEIGHTS = {
    SKILLS: 0.40,
    CAREER: 0.20,
    LANGUAGE: 0.10,
    EXPERIENCE: 0.15,
    BIO: 0.10,
    AVAILABILITY: 0.05
};

const EXPERIENCE_MAP = {
    "Beginner": 1,
    "Entry": 1,
    "Junior": 2,
    "Mid": 3,
    "Mid-level": 3,
    "Senior": 4,
    "Expert": 5,
    "Expert / Lead": 5
};

// Semantic mapping for common skills to handle partial/related matches
const SKILL_MAP = {
    "react": ["frontend", "javascript", "web", "ui", "nextjs", "typescript"],
    "node": ["backend", "javascript", "server", "api", "express", "sql"],
    "python": ["data science", "backend", "ai", "ml", "django", "flask"],
    "design": ["ui", "ux", "product", "figma", "visual", "prototyping"],
    "management": ["leadership", "strategy", "product", "agile", "scrum", "business"],
    "aws": ["cloud", "devops", "infrastructure", "azure", "gcp", "docker"],
    "javascript": ["frontend", "web", "logic", "typescript", "es6"],
    "marketing": ["growth", "seo", "content", "strategy", "social media"],
    "sales": ["business development", "negotiation", "strategy", "crm"]
};

/**
 * Normalizes a list of skills into a semantic set.
 */
const getSemanticSkills = (skills = []) => {
    const semanticSet = new Set();
    skills.forEach(skill => {
        const s = skill.toLowerCase().trim();
        if (!s) return;
        semanticSet.add(s);
        if (SKILL_MAP[s]) {
            SKILL_MAP[s].forEach(related => semanticSet.add(related));
        }
    });
    return semanticSet;
};

/**
 * Calculates Cosine Similarity between two sets (simplified for feature overlap).
 */
const calculateSimilarity = (setA, setB) => {
    if (setA.size === 0 || setB.size === 0) return 0;
    const intersection = new Set([...setA].filter(x => setB.has(x)));
    return intersection.size / Math.sqrt(setA.size * setB.size);
};

/**
 * Calculates language overlap score.
 */
const calculateLanguageScore = (userLangs = [], targetLangs = []) => {
    const userSet = new Set(userLangs.map(l => l.toLowerCase().trim()));
    const targetSet = new Set(targetLangs.map(l => l.toLowerCase().trim()));
    const intersection = [...userSet].filter(l => targetSet.has(l));
    return intersection.length > 0 ? 1 : 0;
};

/**
 * Calculates bio keyword overlap (lightweight TF-IDF substitute).
 */
const calculateBioScore = (userBio = "", targetBio = "") => {
    if (!userBio || !targetBio) return 0;
    const userWords = new Set(userBio.toLowerCase().split(/\W+/).filter(w => w.length > 3));
    const targetWords = new Set(targetBio.toLowerCase().split(/\W+/).filter(w => w.length > 3));
    if (userWords.size === 0) return 0;
    const intersection = [...userWords].filter(w => targetWords.has(w));
    return intersection.length / userWords.size;
};

/**
 * Calculates a compatibility score from 0-100 between two profiles.
 */
export const calculateCompatibility = (user, target) => {
    if (!user || !target || user.role === target.role) return 0;

    let finalScore = 0;

    // 1. Semantic Skill Match (40%) - Updated Weight
    const userSkills = getSemanticSkills(user.skills || user.interests || []);
    const targetSkills = getSemanticSkills(target.skills || target.expertiseAreas || []);
    const skillSim = calculateSimilarity(userSkills, targetSkills);
    finalScore += skillSim * 100 * WEIGHTS.SKILLS;

    // 2. Career Alignment (20%) - Updated Weight
    const userCareer = user.career?.toLowerCase() || "";
    const targetCareer = target.career?.toLowerCase() || "";
    const userCareerWords = new Set(userCareer.split(/[,\s]+/).filter(w => w.length > 2));
    const targetCareerWords = new Set(targetCareer.split(/[,\s]+/).filter(w => w.length > 2));
    const careerSim = calculateSimilarity(userCareerWords, targetCareerWords);
    finalScore += careerSim * 100 * WEIGHTS.CAREER;

    // 3. Language Match (10%) - New Category
    const langScore = calculateLanguageScore(user.languages, target.languages);
    finalScore += langScore * 100 * WEIGHTS.LANGUAGE;

    // 4. Experience Delta (15%)
    const userExp = EXPERIENCE_MAP[user.experienceLevel] || 1;
    const targetExp = EXPERIENCE_MAP[target.experienceLevel] || 1;
    const expDiff = targetExp - userExp;

    let expScore = 0;
    if (expDiff > 1) expScore = 1;
    else if (expDiff === 1) expScore = 0.5;
    else expScore = 0;

    finalScore += expScore * 100 * WEIGHTS.EXPERIENCE;

    // 5. Bio Keyword Match (10%) - New Category
    const bioScore = calculateBioScore(user.bio, target.bio);
    finalScore += bioScore * 100 * WEIGHTS.BIO;

    // 6. Availability Match (5%)
    const userSchedule = user.availability?.schedule || [];
    const targetSchedule = target.availability?.schedule || [];
    const userDays = new Set(userSchedule.filter(d => d.available).map(d => d.day));
    const targetDays = new Set(targetSchedule.filter(d => d.available).map(d => d.day));

    if (userDays.size > 0 && targetDays.size > 0) {
        const matchingDays = [...userDays].filter(day => targetDays.has(day));
        if (matchingDays.length > 0) {
            finalScore += 100 * WEIGHTS.AVAILABILITY;
        }
    }

    // BONUS: Mentorship Style Match
    const userStyle = (user.mentoringStyle || user.mentorshipStyle || "").toLowerCase();
    const targetStyle = (target.mentoringStyle || target.mentorshipStyle || "").toLowerCase();
    if (userStyle && targetStyle && userStyle === targetStyle) {
        finalScore += 15;
    }

    // BONUS: Country Preference
    const preferredCountries = user.preferredCountries || [];
    if (preferredCountries.length > 0 && target.country && preferredCountries.includes(target.country)) {
        finalScore += 10;
    }

    return Math.min(Math.round(finalScore), 100);
};

/**
 * Ranks a list of users based on compatibility with current user.
 */
export const rankMatches = (currentUser, otherUsers) => {
    if (!currentUser) return [];

    const ranked = otherUsers
        .filter(u => u.id !== currentUser.id && u.role !== currentUser.role)
        .map(u => ({
            ...u,
            compatibility: calculateCompatibility(currentUser, u)
        }))
        .sort((a, b) => b.compatibility - a.compatibility);

    // Mentees see top 10 mentors (increased from 5)
    if (currentUser.role === 'mentee') {
        return ranked.slice(0, 10);
    }

    return ranked;
};

File: matching_engine.py
"""
AI Matching Engine - Vector Space Model (Python)
Calculates compatibility scores using feature vectorization and cosine similarity.
"""

import math
from typing import List, Dict, Set, Optional

WEIGHTS = {
    "SKILLS": 0.45,
    "CAREER": 0.35,
    "EXPERIENCE": 0.15,
    "AVAILABILITY": 0.05
}

EXPERIENCE_MAP = {
    "Entry": 1,
    "Mid-level": 2,
    "Senior": 3,
    "Expert / Lead": 4
}

# Semantic mapping for common skills to handle partial/related matches
SKILL_MAP = {
    "react": ["frontend", "javascript", "web", "ui"],
    "node": ["backend", "javascript", "server", "api"],
    "python": ["data science", "backend", "ai", "ml"],
    "design": ["ui", "ux", "product", "figma"],
    "management": ["leadership", "strategy", "product", "agile"],
    "aws": ["cloud", "devops", "infrastructure"],
    "javascript": ["frontend", "web", "logic"]
}


def get_semantic_skills(skills: Optional[List[str]] = None) -> Set[str]:
    """
    Normalizes a list of skills into a semantic set.
    
    Args:
        skills: List of skill strings
        
    Returns:
        Set of normalized skills including semantic relationships
    """
    if skills is None:
        skills = []
    
    semantic_set = set()
    for skill in skills:
        s = skill.lower().strip()
        semantic_set.add(s)
        if s in SKILL_MAP:
            for related in SKILL_MAP[s]:
                semantic_set.add(related)
    
    return semantic_set


def calculate_similarity(set_a: Set[str], set_b: Set[str]) -> float:
    """
    Calculates Cosine Similarity between two sets (simplified for feature overlap).
    
    Args:
        set_a: First set of features
        set_b: Second set of features
        
    Returns:
        Similarity score between 0 and 1
    """
    if len(set_a) == 0 or len(set_b) == 0:
        return 0.0
    
    intersection = set_a & set_b
    denominator = math.sqrt(len(set_a) * len(set_b))
    
    return len(intersection) / denominator


def calculate_compatibility(user: Dict, target: Dict) -> int:
    """
    Calculates a compatibility score from 0-100 between two profiles.
    
    Args:
        user: The current logged-in user profile
        target: The profile being evaluated
        
    Returns:
        Compatibility score (0-100)
    """
    if not user or not target or user.get("role") == target.get("role"):
        return 0
    
    final_score = 0.0
    
    # 1. Semantic Skill Match (45%)
    user_skills = get_semantic_skills(user.get("skills", []))
    target_skills = get_semantic_skills(target.get("skills", []))
    skill_sim = calculate_similarity(user_skills, target_skills)
    final_score += skill_sim * 100 * WEIGHTS["SKILLS"]
    
    # 2. Career Alignment (35%)
    user_career = (user.get("career") or "").lower()
    target_career = (target.get("career") or "").lower()
    
    # Check for keyword overlaps in career description
    user_career_words = set(word for word in user_career.split() if len(word) > 2)
    target_career_words = set(word for word in target_career.split() if len(word) > 2)
    career_sim = calculate_similarity(user_career_words, target_career_words)
    
    # Explicit role goal check for mentees
    if user.get("role") == "mentee":
        if user_career in target_career:
            final_score += 20 * WEIGHTS["CAREER"]
    
    final_score += career_sim * 100 * WEIGHTS["CAREER"]
    
    # 3. Experience Delta (15%)
    user_exp = EXPERIENCE_MAP.get(user.get("experienceLevel"), 1)
    target_exp = EXPERIENCE_MAP.get(target.get("experienceLevel"), 1)
    
    if user.get("role") == "mentee":
        # Mentees want more experienced mentors
        if target_exp > user_exp:
            final_score += 100 * WEIGHTS["EXPERIENCE"]
        elif target_exp == user_exp:
            final_score += 50 * WEIGHTS["EXPERIENCE"]
    else:
        # Mentors ideally match with less experienced mentees
        if target_exp < user_exp:
            final_score += 100 * WEIGHTS["EXPERIENCE"]
        elif target_exp == user_exp:
            final_score += 50 * WEIGHTS["EXPERIENCE"]
    
    # 4. Availability Match (5%)
    if user.get("availability") == target.get("availability"):
        final_score += 100 * WEIGHTS["AVAILABILITY"]
    
    return min(round(final_score), 100)


def rank_matches(current_user: Dict, other_users: List[Dict]) -> List[Dict]:
    """
    Ranks a list of users based on compatibility with current user.
    
    Args:
        current_user: The current user's profile
        other_users: List of all user profiles from database
        
    Returns:
        List of profiles ranked by compatibility score
    """
    if not current_user:
        return []
    
    current_user_id = current_user.get("id")
    current_user_role = current_user.get("role")
    
    # Filter and score other users
    ranked = []
    for user in other_users:
        if user.get("id") != current_user_id and user.get("role") != current_user_role:
            scored_user = user.copy()
            scored_user["compatibility"] = calculate_compatibility(current_user, user)
            ranked.append(scored_user)
    
    # Sort by compatibility (descending)
    ranked.sort(key=lambda x: x["compatibility"], reverse=True)
    
    # Mentees see top 5 mentors, Mentors see all (but ranked)
    if current_user_role == "mentee":
        return ranked[:5]
    
    return ranked


# Example usage / Testing
if __name__ == "__main__":
    # Test data
    mentor = {
        "id": "1",
        "role": "mentor",
        "skills": ["React", "Python", "AWS"],
        "career": "Full-stack software engineer",
        "experienceLevel": "Senior",
        "availability": "Weekends"
    }
    
    mentee = {
        "id": "2",
        "role": "mentee",
        "skills": ["JavaScript", "Python"],
        "career": "Backend developer",
        "experienceLevel": "Entry",
        "availability": "Weekends"
    }
    
    compatibility = calculate_compatibility(mentee, mentor)
    print(f"Compatibility Score: {compatibility}/100")
    
    # Test ranking
    all_users = [mentor, mentee]
    ranked = rank_matches(mentee, all_users)
    print(f"\nRanked Matches for mentee:")
    for match in ranked:
        print(f"  - {match.get('name', 'User')}: {match['compatibility']}/100")
